control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -4498,187 +4469,285 @@ func healthCheckIdentityProvider(ctx context.Context, hcp *hyperv1.HostedControl
+	// the last reconcile was done.
+	cond := meta.FindStatusCondition(hcp.Status.Conditions, string(hyperv1.AWSSecurityGroupsReconciled))
+	cond.LastTransitionTime = securityGroupsReconciled.LastTransitionTime
 
 	if err := r.Client.Status().Patch(ctx, hcp, client.MergeFromWithOptions(originalHCP, client.MergeFromWithOptimisticLock{})); err != nil {
-		return fmt.Errorf("failed to update status: %w", err)
+		return fmt.Errorf("failed to update status for AWS security groups: %w", err)
 	}
-
-	return creationErr
+	return nil
 }
 
-func awsSecurityGroupFilters(infraID string) []*ec2.Filter {
+func awsSecurityGroupFilters(infraID, name string) []*ec2.Filter {
 	return []*ec2.Filter{
 		{
 			Name:   awssdk.String(fmt.Sprintf("tag:kubernetes.io/cluster/%s", infraID)),
 			Values: []*string{awssdk.String("owned")},
 		},
 		{
 			Name:   awssdk.String("tag:Name"),
-			Values: []*string{awssdk.String(awsSecurityGroupName(infraID))},
+			Values: []*string{awssdk.String(name)},
 		},
 	}
 }
 
-func awsSecurityGroupName(infraID string) string {
+func awsDefaultSecurityGroupName(infraID string) string {
 	return fmt.Sprintf("%s-default-sg", infraID)
 }
 
-func createAWSDefaultSecurityGroup(ctx context.Context, ec2Client ec2iface.EC2API, infraID, vpcID string, additionalTags []hyperv1.AWSResourceTag) (string, error) {
+func awsVPCEndpointSecurityGroupName(infraID string) string {
+	return fmt.Sprintf("%s-vpc-endpoint-sg", infraID)
+}
+
+func (r *HostedControlPlaneReconciler) ensureAWSSecurityGroups(ctx context.Context, hcp *hyperv1.HostedControlPlane) (string, string, error) {
 	logger := ctrl.LoggerFrom(ctx)
 
 	// Determine VPC cidr
-	vpcResult, err := ec2Client.DescribeVpcsWithContext(ctx, &ec2.DescribeVpcsInput{
-		VpcIds: []*string{awssdk.String(vpcID)},
-	})
+	vpcID := hcp.Spec.Platform.AWS.CloudProviderConfig.VPC
+	shouldReconcileVPCEndpointSG := util.IsPrivateHCP(hcp)
+	defaultSG, err := existingSG(ctx, r.ec2Client, awsDefaultSecurityGroupName(hcp.Spec.InfraID), hcp.Spec.InfraID)
 	if err != nil {
-		logger.Error(err, "Failed to describe vpc", "vpcID", vpcID)
-		return "", fmt.Errorf("failed to describe vpc %s, code %s", vpcID, awsErrorCode(err))
+		return "", "", err
 	}
-	if len(vpcResult.Vpcs) == 0 {
-		return "", fmt.Errorf("vpc %s not found", vpcID)
+	var vpcEndpointSG *ec2.SecurityGroup
+	if shouldReconcileVPCEndpointSG {
+		vpcEndpointSG, err = existingSG(ctx, r.ec2Client, awsVPCEndpointSecurityGroupName(hcp.Spec.InfraID), hcp.Spec.InfraID)
+		if err != nil {
+			return "", "", err
+		}
 	}
-	vpcCIDR := awssdk.StringValue(vpcResult.Vpcs[0].CidrBlock)
-	describeSGResult, err := ec2Client.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{Filters: awsSecurityGroupFilters(infraID)})
+	if defaultSG == nil {
+		defaultSG, err = createAWSSecurityGroup(ctx, r.ec2Client, hcp.Spec.InfraID, vpcID, awsDefaultSecurityGroupName(hcp.Spec.InfraID), "default worker security group", hcp.Spec.Platform.AWS.ResourceTags)
```

Comment:
nit: these functions already take in the infra ID, perhaps the SG name should be a `func(id) string` to ensure the expected relationship is invariant?